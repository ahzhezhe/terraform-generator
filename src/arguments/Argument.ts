import { TerraformElement, Util } from '../utils';

/**
 * @category Argument
 */
export class Argument<T extends string = string> extends TerraformElement {

  readonly #argument: T;

  /**
   * Construct argument.
   *
   * @param arg argument as string
   */
  constructor(arg: T) {
    super();
    this.#argument = arg;
  }

  /**
   * Get argument's attribute.
   *
   * @param name attribute name
   */
  attr(name: string): Argument {
    return new Argument(`${this.#argument}.${name.trim()}`);
  }

  /**
   * Get list argument's element.
   *
   * @param idx element index
   */
  element(idx: number): Argument {
    return new Argument(`${this.#argument}[${idx}]`);
  }

  /**
   * To Terraform representation.
   *
   * Use this method when argument is used as an interpolation in another Terraform argument or code.
   */
  override toTerraform(): string {
    return Util.escape(this.#argument);
  }

  /**
   * To string.
   *
   * Use this method when argument is used as an interpolation in a Terraform string or heredoc.
   *
   * It is automatically called when argument is used in template literal.
   */
  toString(): string {
    return `\${${this.toTerraform()}}`;
  }

}

/**
 * Convenient function to construct new [[Argument]].
 *
 * @param arg argument as string or copy from another argument object
 *
 * @category Argument
 */
export const arg = <T extends string = string>(arg: T): Argument<T> => new Argument(arg);
