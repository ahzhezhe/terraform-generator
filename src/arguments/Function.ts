import { Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Function extends Argument {

  /**
   * Construct function argument.
   *
   * @param name function name
   * @param args function arguments
   */
  constructor(name: string, ...args: any[]) {
    super(Function.#constructArgument(name, ...args));
  }

  static #constructArgument(fn: string, ...args: any[]): string {
    let str = `${fn}(`;
    args.forEach((arg, i) => {
      str += Util.argumentValueToString(arg);
      if (i < args.length - 1) {
        str += ', ';
      }
    });
    str += ')';
    return str;
  }

}

/**
 * Convenient function to construct new [[Function]].
 *
 * @param name function name
 * @param args function arguments
 *
 * @category Argument
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export const fn = (name: string, ...args: any[]): Function => new Function(name, ...args);
