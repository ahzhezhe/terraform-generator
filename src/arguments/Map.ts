import { TerraformArgs, Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Map extends Argument {

  readonly arguments: TerraformArgs;

  /**
   * Construct map.
   *
   * @param args map values
   */
  constructor(args: TerraformArgs) {
    super(Util.argumentValueToString(args)!);
    this.arguments = args;
  }

}

/**
 * Convenient function to construct new [[Map]].
 *
 * @param args map values
 *
 * @category Argument
 */
export const map = (args: TerraformArgs): Map => new Map(args);
