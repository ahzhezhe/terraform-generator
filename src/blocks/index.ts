export * from './Block';
export * from './Backend';
export * from './Comment';
export * from './Data';
export * from './Import';
export * from './Locals';
export * from './Module';
export * from './Moved';
export * from './Output';
export * from './Provider';
export * from './Provisioner';
export * from './Removed';
export * from './Resource';
export * from './Variable';
