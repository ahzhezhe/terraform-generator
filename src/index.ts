export { TerraformElement, TerraformArgs } from './utils';

export { TerraformGenerator, WriteOptions } from './TerraformGenerator';

export {
  Block,
  Backend,
  Comment,
  Data,
  Import, ImportArgs,
  Locals,
  Module, ModuleArgs,
  Moved, MovedArgs,
  Output, OutputArgs,
  Provider,
  Provisioner, FileProvisionerArgs, LocalExecProvisionerArgs, RemoteExecProvisionerArgs,
  Removed, RemovedArgs,
  Resource, ResourceToDataOptions,
  Variable, VariableArgs
} from './blocks';

export {
  Argument, arg,
  Attribute, attr,
  Function, fn,
  Heredoc, heredoc,
  List, list,
  Map, map
} from './arguments';
