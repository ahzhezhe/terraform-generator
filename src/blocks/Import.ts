import { Argument, Attribute } from '../arguments';
import { TerraformElement, Util } from '../utils';
import { Block } from '.';

/**
 * @category Block
 */
export interface ImportArgs {
  to: TerraformElement;
  id: TerraformElement | string;
  provider?: Argument;
}

/**
 * @category Block
 */
export class Import extends Block<ImportArgs> {

  /**
   * Construct import.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  constructor(args: ImportArgs) {
    super('import', [], args);
  }

  override asArgument(): Argument {
    throw Util.inaccessibleMethod();
  }

  override attr(_name: string): Attribute {
    throw Util.inaccessibleMethod();
  }

}
