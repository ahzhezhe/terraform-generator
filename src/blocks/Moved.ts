import { Argument, Attribute } from '../arguments';
import { TerraformElement, Util } from '../utils';
import { Block } from '.';

/**
 * @category Block
 */
export interface MovedArgs {
  from: TerraformElement;
  to: TerraformElement;
}

/**
 * @category Block
 */
export class Moved extends Block<MovedArgs> {

  /**
   * Construct moved.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  constructor(args: MovedArgs) {
    super('moved', [], args);
  }

  override asArgument(): Argument {
    throw Util.inaccessibleMethod();
  }

  override attr(_name: string): Attribute {
    throw Util.inaccessibleMethod();
  }

}
