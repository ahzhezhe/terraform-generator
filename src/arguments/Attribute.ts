import { Block } from '../blocks';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Attribute extends Argument {

  /**
   * Construct block's attribute.
   *
   * @param block block
   * @param attrName attribute name
   */
  constructor(block: Block, attrName: string) {
    super(`${block.asArgument().toTerraform()}.${attrName.trim()}`);
  }

}

/**
 * Convenient function to construct new block's [[Attribute]].
 *
 * @param block block
 * @param name attribute name
 *
 * @category Argument
 */
export const attr = (block: Block, name: string): Attribute => new Attribute(block, name);
