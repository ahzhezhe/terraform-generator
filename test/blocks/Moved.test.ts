import { arg } from '../../src/arguments';
import { Moved, Resource } from '../../src/blocks';

test('To argument', () => {
  const moved = new Moved({
    from: arg('resource.a'),
    to: arg('resource.b')
  });
  expect(moved.toTerraform()).toMatchSnapshot();
  expect(() => moved.asArgument()).toThrow();
  expect(() => moved.attr('attr')).toThrow();
});

test('To resource', () => {
  const resource = new Resource('resource', 'b');
  const moved = new Moved({
    from: arg('resource.a'),
    to: resource
  });
  expect(moved.toTerraform()).toMatchSnapshot();
  expect(() => moved.asArgument()).toThrow();
  expect(() => moved.attr('attr')).toThrow();
});
