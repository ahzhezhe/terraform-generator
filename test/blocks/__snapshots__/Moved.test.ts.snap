// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`To argument 1`] = `
"moved{
from = resource.a
to = resource.b
}

"
`;

exports[`To resource 1`] = `
"moved{
from = resource.a
to = resource.b
}

"
`;
